fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/JoinEquivalentPredicatePushDown.java

```diff
@@ -0,0 +1,136 @@
+            this.columnRefFactory = columnRefFactory;
+        }
+
+        public Void visit(OptExpression optExpression, RewriteContext context) {
+            //优先遍历子孩子
+            for (OptExpression o : optExpression.getInputs()) {
+                o.getOp().accept(this, o, context);
+            }
+
+            ScalarOperator predicate = optExpression.getOp().getPredicate();
+            if (predicate != null) {
+                context.columnRefToConstant.addAll(Utils.extractConjuncts(predicate));
+            }
+
+            //LogicalAnchor
+            if (optExpression.getOutputColumns() != null) {
+                Iterator<ScalarOperator> iter = context.columnRefToConstant.iterator();
+                while (iter.hasNext()) {
+                    ScalarOperator entry = iter.next();
+                    if (!optExpression.getOutputColumns().containsAll(entry.getUsedColumns())) {
+                        iter.remove();
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitLogicalJoin(OptExpression optExpression, RewriteContext context) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) optExpression.getOp();
+
+            List<ScalarOperator> pushDownToLeft = new ArrayList<>();
+            List<ScalarOperator> pushDownToRight = new ArrayList<>();
+
+            if (joinOperator.getJoinType().isFullOuterJoin()) {
+                return null;
+            }
+
+            if (!joinOperator.getJoinType().isRightOuterJoin() && !joinOperator.getJoinType().isRightAntiJoin()) {
+                getPredicatePushDownToRight(optExpression, joinOperator, pushDownToRight);
+            }
+
+            if (!joinOperator.getJoinType().isLeftOuterJoin() && !joinOperator.getJoinType().isLeftAntiJoin()) {
+                getPredicatePushDownToLeft(optExpression, joinOperator, pushDownToLeft);
+            }
+
+            PushDownJoinPredicateBase.pushDownPredicate(optExpression, pushDownToLeft, pushDownToRight);
+            return null;
+        }
+
+        void getPredicatePushDownToRight(OptExpression optExpression, LogicalJoinOperator joinOperator,
+                                         List<ScalarOperator> pushDownToRight) {
+            RewriteContext rleft = new RewriteContext();
+            optExpression.inputAt(0).getOp().accept(this, optExpression.inputAt(0), rleft);
+
+            ScalarEquivalenceExtractor scalarEquivalenceExtractor = new ScalarEquivalenceExtractor();
+            List<ScalarOperator> inputPredicates = Utils.extractConjuncts(joinOperator.getOnPredicate());
+            scalarEquivalenceExtractor.union(inputPredicates);
+            for (ScalarOperator entry : rleft.columnRefToConstant) {
+                scalarEquivalenceExtractor.union(Utils.extractConjuncts(entry));
+            }
+
+            for (int columnId : optExpression.getInputs().get(1).getOutputColumns().getColumnIds()) {
```

Comment:
Is this can support sql like `select * from t0 left join t1 on t0.v1 = t1.v1 where t0.v1 < 10` ?